Add WebView JavaScript channels (Dart side).

amirh opened the pull request:
Allows JavaScript code running in the WebView to send messages that will be received by the Flutter app's Dart code.

To keep the PRs smaller this does not include the platform side implementations which will be sent in following PRs.

Right now there are no return values on the JavaScript side, we could think of returning a JavaScript promise (or a promise equivalent) in the future, but as a first step I start without it (`evaluateJavascript` can be used to pass messages in the other direction so this shouldn't be a blocking limitation).
I believe that if we end up adding a return value on the JavaScript side this could be done without making a breaking change.

### Sample usage
Dart:
```dart
WebView(
  javascriptChannels: <JavascriptChannel>[
    JavascriptChannel(name: 'Print', onMessageReceived: (String msg) {print(msg); }),
  ].toSet(),
),
```

JavaScript (running inside the WebView):
```javascript
Print.postMessage('Hello!');
```


Fixes: https://github.com/flutter/flutter/issues/24837

mklim commented on packages/webview_flutter/lib/webview_flutter.dart:
````diff
@@ -56,13 +83,39 @@ class WebView extends StatefulWidget {
+  /// JavaScript code can call:
+  ///
+  /// ```javascript
+  /// Print.postMessage('Hello');
````
I think it would be worth establishing here how namespace collision is handled on the JS side. What happens if somebody tries to name their channel `Array`?

amirh replied:
Discussed offline, we're going to override whatever window property that is there as this is what Android is doing. We don't think it's a security risk as the app developer has the power to do something like this anyway by injecting JavaScript.

amirh commented on packages/webview_flutter/lib/webview_flutter.dart:
```diff
@@ -19,6 +19,32 @@ enum JavascriptMode {
+class JavascriptChannel {
+  /// Constructs a Javascript channel.
+  ///
+  /// The parameters `name` and `onMessageReceived` must not be null.
```
Good catch.

I changed it to require that channel names follow the Dart identifier naming rules (only letters digits and underscores allowed, can't start with a digit).

amirh commented on packages/webview_flutter/test/webview_flutter_test.dart:
```diff
@@ -353,6 +346,86 @@ void main() {
       throwsA(anything),
     );
   });
+
```
done

amirh commented:
@mklim PTAL, changes since your last review:
1. Updated documentation: b57117f
2. Bump version and update change log: 59e1248
3. Updated the example app: 9a350dd

amirh commented on packages/webview_flutter/example/lib/main.dart:
```diff
@@ -23,17 +23,34 @@ class WebViewExample extends StatelessWidget {
-        },
-      ),
+      // We're using a Builder here so we have a context that is below the Scaffold
+      // to allow calling Scaffold.maketoast
```
thanks, fixed.

kingctan commented:
is there any way to add callback in js side which can postmessage and get a return object?

yexiaodong commented:
is there any way to add callback in js side which can postmessage and get a return object?
